cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,650 @@
+    seconds                 = simt::std::chrono::floor<duration_s>(duration_ns(duration));
+    timeparts[DU_SUBSECOND] = modulo_time(duration, 1000L * 1000 * 1000);
+  }
+  timeparts[DU_DAY] = simt::std::chrono::floor<duration_D>(seconds).count();
+  timeparts[DU_HOUR] =
+    modulo_time(simt::std::chrono::floor<simt::std::chrono::hours>(seconds).count(), 24);
+  timeparts[DU_MINUTE] =
+    modulo_time(simt::std::chrono::floor<simt::std::chrono::minutes>(seconds).count(), 60);
+  timeparts[DU_SECOND] = modulo_time(seconds.count(), 60);
+}
+
+template <typename T>
+struct duration_to_string_size_fn {
+  const column_device_view d_durations;
+  const format_item* d_format_items;
+  size_type items_count;
+  type_id type;
+
+  __device__ int8_t count_trailing_zeros(int n) const
+  {
+    int8_t zeros = 0;
+    if ((n % 100000000) == 0) {
+      zeros += 8;
+      n /= 100000000;
+    }
+    if ((n % 10000) == 0) {
+      zeros += 4;
+      n /= 10000;
+    }
+    if ((n % 100) == 0) {
+      zeros += 2;
+      n /= 100;
+    }
+    if ((n % 10) == 0) { zeros++; }
+    return zeros;
+  }
+
+  __device__ int8_t format_length(char format_char, int32_t const* const timeparts) const
+  {
+    switch (format_char) {
+      case 'd': return count_digits(timeparts[DU_DAY]); break;
+      case '+': return timeparts[DU_DAY] < 0 ? 1 : 0; break;
+      case 'H': return count_digits(timeparts[DU_HOUR]); break;
+      case 'M': return count_digits(timeparts[DU_MINUTE]); break;
+      case 'S': return count_digits(timeparts[DU_SECOND]); break;
+      // 0 or 6 digits for pandas, include dot only if non-zero.
+      case 'u': return (timeparts[DU_SUBSECOND] == 0) ? 0 : 6 + 1; break;
+      // 0 or ns without trailing zeros
+      case 'f':
+        return (timeparts[DU_SUBSECOND] == 0) ? 0 : [units = type] {
+          if (units == type_id::DURATION_MILLISECONDS) return 3 + 1;  // +1 is for dot
+          if (units == type_id::DURATION_MICROSECONDS) return 6 + 1;  // +1 is for dot
+          if (units == type_id::DURATION_NANOSECONDS) return 9 + 1;   // +1 is for dot
+          return 0;
+        }() - count_trailing_zeros(timeparts[DU_SUBSECOND]);  // 3/6/9-trailing_zeros.
+        break;
+      default: return 2;
+    }
+  }
+
+  __device__ size_type operator()(size_type idx)
+  {
+    if (d_durations.is_null(idx)) return 0;
+    auto duration                   = d_durations.element<T>(idx);
+    int32_t timeparts[DU_ARRAYSIZE] = {0};  // days, hours, minutes, seconds, subseconds(9)
+    dissect_duration(duration.count(), timeparts, type);
+    size_type string_length{0};
+    for (auto i = 0; i < items_count; i++) {
+      format_item item = d_format_items[i];
+      if (item.item_type == format_char_type::literal)
+        string_length++;
+      else if (item.length != -1)
+        string_length += item.length;
+      else
+        string_length += format_length(item.value, timeparts);
+    }
+    return string_length;
```

Comment:
Not sure if the above comment will render my comment obsolete but this is a `std::accumulate`

Reply:
done. used `thrust::transform_reduce`